app/sign.js

```diff
@@ -0,0 +1,356 @@
+  backupKeyNode.keyPair.network = network;
+
+  const txBuilder = utxoLib.TransactionBuilder.fromTransaction(transaction, network);
+
+  _.forEach(recoveryRequest.inputs, function(input, i) {
+    const isBech32 = !input.redeemScript;
+    const isSegwit = !!input.witnessScript;
+
+    // chain paths come from the SDK with a leading /, which is technically not allowed by BIP32
+    if (input.chainPath.startsWith('/')) {
+      input.chainPath = input.chainPath.slice(1);
+    }
+
+    const derivedHDNode = backupKeyNode.derivePath(input.chainPath);
+
+    console.log(`Signing input ${ i + 1 } of ${ recoveryRequest.inputs.length } with ${ derivedHDNode.neutered().toBase58() } (${ input.chainPath })`);
+
+    if (isBech32) {
+      const witnessScript = Buffer.from(input.witnessScript, 'hex');
+      const witnessScriptHash = utxoLib.crypto.sha256(witnessScript);
+      const prevOutScript = utxoLib.script.witnessScriptHash.output.encode(witnessScriptHash);
+      txBuilder.sign(i, derivedHDNode.keyPair, prevOutScript, utxoLib.Transaction.SIGHASH_ALL, input.amount, witnessScript);
+    } else {
+      const redeemScript = new Buffer(input.redeemScript, 'hex');
+
+      if (isSegwit) {
+        const witnessScript = new Buffer(input.witnessScript, 'hex');
+        txBuilder.sign(i, derivedHDNode.keyPair, redeemScript, utxoLib.Transaction.SIGHASH_ALL, input.amount, witnessScript)
+      } else {
+        txBuilder.sign(i, derivedHDNode.keyPair, redeemScript, utxoLib.Transaction.SIGHASH_ALL);
+      }
+    }
+  });
+
+  return txBuilder.build().toHex();
+};
+
+const handleSignEthereum = function(recoveryRequest, key, skipConfirm) {
+  const EthTx = require('ethereumjs-tx');
+
+  const transaction = new EthTx(recoveryRequest.tx);
+  const decimals = coinDecimals[recoveryRequest.coin];
+
+  const customMessage = recoveryRequest.custom ? recoveryRequest.custom.message : 'None';
+  const txData = transaction.data;
+  const outputs = [{
+    address: '0x' + txData.slice(16, 36).toString('hex'),
+    amount: (new BN(txData.slice(36, 68).toString('hex'), 16)).div(TEN.pow(decimals))
+  }];
+
+  confirmRecovery(recoveryRequest.backupKey, outputs, customMessage, skipConfirm);
+
+  if (!key) {
+    console.log('Please enter the xprv of the wallet for signing: ');
+    key = prompt();
+  }
+
+  let backupKeyNode;
+
+  try {
+    backupKeyNode = utxoLib.HDNode.fromBase58(key);
+  } catch (e) {
+    throw new Error('invalid private key');
+  }
+
+  if (backupKeyNode.toBase58() === backupKeyNode.neutered().toBase58()) {
+    throw new Error('please provide the private (not public) wallet key');
+  }
+
+  if (backupKeyNode.neutered().toBase58() !== recoveryRequest.backupKey) {
+    throw new Error('provided private key does not match public key specified with recovery request');
+  }
+
+  const backupHDNode = prova.HDNode.fromBase58(key);
+  const backupSigningKey = backupHDNode.getKey().getPrivateKeyBuffer();
+
+  transaction.sign(backupSigningKey);
```

Comment:
Are we just leaving it up to the KRS to build the ETH function call to the contract and stuff before signing? Separately, as in BGJS/eth.js:264-5, they create the operationHash specifically with the contract function call data and then signs that. And it looks like ethereumjs-tx computes the hash differently

Reply:
KRS doesn't build the function call - the user does through the WRW, or BitGo does through bgadmin.